### DISCUSSION TITLE
Another one about indirect relations...

### DISCUSSION THREAD

**Discussion by joeinnes** [pocketbase/pocketbase#2282] [Q&A] [answered]
Hi,

Hopefully this will make sense.

I have four tables:

- Recipes
- Ingredients
- PlannedMeals
- Recipe_Ingredients

`Recipe_Ingredients` holds a reference to the `Recipe`, a reference to the `Ingredient`, plus a quantity (which is why this isn't a simple relation).

Now I want to craft a query which will allow me to build a shopping list. 

However, using the current indirect expansion syntax of `Table(ReferenceField)` does not match anything, because the `ReferenceField` assumes that the reference is to the current table's `id`, as far as I can tell.

I am not familiar with Go, but I *think* what I'm trying to understand is on the following line. If I understand correctly, the **ID** of the current record is taken as the foreign key on the indirectly expanded relation here.

https://github.com/pocketbase/pocketbase/blob/3ea02c945dc150c29d4f448109e2656b23ec398e/daos/record_expand.go#L98

My questions are twofold:

1. Can you confirm that I am correct in my understanding that currently, it is not possible to use any field *other* than the ID to query an indirectly expanded reference?
2. If I create my own, patched version of PocketBase which allows looking up by another field, would you be interested in a PR (bearing in mind I have no Go experience), and if so, is there any particular syntax you would suggest? Currently, I'm thinking something along the lines of `RecipeIngredients(@meal=meal)`, which can be split into `@meal` and `meal`, where `@meal` means 'look for the current record's `meal` value', and the second part refers to the foreign key on the indirectly related record.

**Comment by ganigeorgiev** [comment]
> 1. Can you confirm that I am correct in my understanding that currently, it is not possible to use any field other than the ID to query an indirectly expanded reference?

Yes. Only `relation` fields are used to construct the indirect expand reference. If you haven't already, for an example you can check https://pocketbase.io/docs/expanding-relations/#indirect-expand.

> 2. If I create my own, patched version of PocketBase which allows looking up by another field, would you be interested in a PR.

For now there are no plans to add support for other fields in the indirect expand matching since the behavior could be ambiguous (eg. the above could ended up matching another record, not the current main one).
If you want to apply a filter-like condition on the indirect expanded results, it might be better to just send a separate request.

**Reply by joeinnes** [reply]
You might be right about sending separate requests, but then it would potentially be quite a lot.

If I have a week of planned meals, each day having two meals, that's 14 recipes, which is 14 queries to run, but if that's the best approach then that's what I'll do.

I think I may have expressed myself badly by naming columns poorly, but there are relation fields linking all the records.

I see what you mean about the ambiguous behaviour, but that's sort of what I want—the 'match' *should* be an ID from a different collection, in this case the 'Recipe' collection (which I poorly named 'meal' in the PlannedMeals table).

Otherwise I will need do something like (need to wrap my head around the async logic here but just as an illustration):

```js
const { items } = await db.collection('PlannedMeals').getFullList();
let queue = items.map(async (item) => {
  let ingredients = await db.collection('Recipes').getOne(item.recipe_id, { expand: 'RecipeIngredients(recipe_id),RecipeIngredients(recipe_id).Ingredients'});
  return {
    ...item,
    ingredients
  };
});
const PlannedMealsWithIngredients = Promise.all(queue);
```

```dbml
Table PlannedMeals {
  id integer [primary key]
  recipe_id integer
}

Table Recipes {
  id integer [primary key]
}

Table Ingredients {
  id integer [primary key]
}

Table RecipeIngredients {
  id integer [primary key]
  ingredient_id integer
  recipe_id integer
}


Ref: PlannedMeals.recipe_id > Recipes.id
Ref: RecipeIngredients.recipe_id > Recipes.id
Ref: RecipeIngredients.ingredient_id > Ingredients.id

```
![Screenshot 2023-04-13 at 13 42 40](https://user-images.githubusercontent.com/4704069/231748271-12913af7-36a3-459b-a3e1-02b73623e7f6.png)

**Reply by ganigeorgiev** [reply, marked as answer]
> If I have a week of planned meals, each day having two meals, that's 14 recipes, which is 14 queries to run, but if that's the best approach then that's what I'll do.

You can perform the grouping on the client side, aka. first extract the ids of the relations you want to expand and then send one request with all ids combined with `||`, something like:

```js
const ids = ["id1", "id2", ...];

const filter = ids.map((id) => `id="${id}"`).join("||");

const records = await pb.collection("example").getFullList({ filter });
```

> Otherwise I will need do something like (need to wrap my head around the async logic here but just as an illustration):

I'm not sure that I understand what is the expected behavior of the provided example. If you want to expand all `Ingredients` from `PlannedMeals`, then you can do something like:

```js
const meals = await pb.collection("PlannedMeals").getFullList({
    expand: "recipe_id.RecipeIngedients(recipe_id).ingredient_id"
});
```

**Reply by joeinnes** [reply]
That's the part I was missing—if I do `recipe_id.` first! That works perfectly. Thank you!
